### ISSUE TITLE
Alternate version matching with string rather than int to handle 0-prefixed numbers (failed ghostscript version check since version 10.01.0)

### ISSUE THREAD

**Issue by Jehan** [flathub-infra/flatpak-external-data-checker#360]
The version checker started to fail for Ghostscript, after release of version [10.01.0](https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/). We use the [anitya](https://release-monitoring.org/project/1157/) `x-data-checker`'s `type` (though this issue might happen with other data checkers, I haven't verified) and here is our [current manifest code for GIMP](https://github.com/flathub/org.gimp.GIMP/blob/d0c65eef8ed30c1fa8bbaf91309c0c8766ded724/org.gimp.GIMP.json#L590-L595): 

```json
                    "x-checker-data": {
                        "type": "anitya",
                        "project-id": 1157,
                        "stable-only": true,
                        "url-template": "https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs$major$minor$patch/ghostscript-$version.tar.gz"
                    }
```

Finally here is an example of [failed build log error](https://buildbot.flathub.org/#/builders/26/builds/3716) because of the version check:

```log
ERROR   src.manifest: Failed to check archive ghostscript/ghostscript-10.01.0.tar.xz with AnityaChecker: Error downloading upstream source: 404, message='Not Found', url=URL('https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs1010/ghostscript-10.01.0.tar.gz')
```

The right URL is: https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs10010/ghostscript-10.01.0.tar.gz

Basically it seems that the substitution code replaces `$minor` by `1` while the exact scheme for ghostscript apparently expects `01` (i.e. taking the version number as a string as-is, keeping the leading `0`, rather than an int), which very likely means that there is a roundtrip to `int` type.

Though it might be the used scheme in some cases, in others (e.g. for ghostscript), it's not. So we'd need alternative substitution variables to handle such case.

Also do you know if there is a workaround for now (other than removing the version check for this module)? Because right now all builds fail on Flathub because of this.

**Comment by Jehan** [comment]
P.S.: checking the code, looking if such alternate variables already exist, I clearly see [the problem](https://github.com/flathub/flatpak-external-data-checker/blob/b3f8e8a0507c8955361bae4d6b709dd08edcc95b/src/checkers/__init__.py#L158-L161):

```py
from distutils.version import LooseVersion
[…]
        version_list = LooseVersion(version).version
[…]
        for i, version_part in enumerate(version_list):
```

And checking how this module works, testing in a console:

```py
In [4]: l = LooseVersion("10.01.0")

In [5]: l.version
Out[5]: [10, 1, 0]
```

So in `version_list`, the leading zeros are already lost.

**Comment by wjt** [comment]
You can use the json checker as a workaround. Try:

```json
"x-checker-data": {
    "type": "json",
    "url": "https://api.github.com/repos/ArtifexSoftware/ghostpdl-downloads/releases/latest",
    "version-query": ".name | split(\" \") | .[1]",
    "url-query": ".assets[] | select(.name|test(\"^ghostscript-(.*).tar.gz$\")) | .browser_download_url"
}

**Comment by Jehan** [comment]
Thanks. Works fine!

**Comment by Alexander-Wilms** [comment]
Maybe one could use f.strings for the url-template and tag-template:

```python
major=10
minor=1
patch=0

print(f"https://github.com/ArtifexSoftware/ghostpdl-downloads/releases/download/gs{major}{minor:02d}{patch}/ghostscript-{version}.tar.gz")
```

Also, since you're looking for an alternative to `distutils.version.looseversion` and the alternative, the `Version` package doesn't seem to support leading zeros either, it might be easiest to parse version numbers by hand and return them as strings to be used in the templates.

Cf.
https://github.com/flathub/flatpak-external-data-checker/issues/355
https://packaging.pypa.io/en/stable/version.html

**Comment by gasinvein** [comment]
f-strings are not an option for untrusted input, since it would open up a very simple way to ACE. But maybe there is another library that can be used for string template (something like jinja2 but simpler)?
